File: test/utils/Signer.ts
import {
  Contract,
  defaultProvider,
  ec,
  encode,
  hash,
  KeyPair,
  number,
  stark,
} from "starknet";

export class Signer {
  public privateKey: string = "";
  public publicKey: string = "";
  public keyPair: KeyPair;

  constructor(privateKey: string) {
    this.privateKey = privateKey;
    this.keyPair = ec.getKeyPair(this.privateKey);
    this.publicKey = ec.getStarkKey(this.keyPair);
  }

  public async sendTransaction(
    account: Contract,
    to: string,
    selectorName: string,
    args: string[]
  ): Promise<string> {
    const { nonce } = await account.call("get_nonce");

    const msgHash = encode.addHexPrefix(
      hash.hashMessage(
        account.connectedTo as string,
        to,
        stark.getSelectorFromName(selectorName),
        args,
        nonce.toString()
      )
    );

    const { r, s } = ec.sign(this.keyPair, msgHash);

    const { code, transaction_hash } = await account.invoke(
      "execute",
      {
        to,
        selector: stark.getSelectorFromName(selectorName),
        calldata: args,
        nonce: nonce.toString(),
      },
      [number.toHex(r), number.toHex(s)]
    );

    await defaultProvider.waitForTx(transaction_hash);

    return code;
  }
}

File: tests/utils/Signer.py
# pylint: disable=invalid-name
from starkware.crypto.signature.signature import private_to_stark_key, sign
from starkware.starknet.public.abi import get_selector_from_name
from starkware.cairo.common.hash_state import compute_hash_on_elements


class Signer():
    def __init__(self, private_key):
        self.private_key = private_key
        self.public_key = private_to_stark_key(private_key)

    def sign(self, message_hash):
        return sign(msg_hash=message_hash, priv_key=self.private_key)

    async def send_transaction(self, account, recipient_address, selector_name, calldata, nonce=None):  # pylint: disable=too-many-arguments
        if nonce is None:
            execution_info = await account.get_nonce().call()
            nonce, = execution_info.result

        selector = get_selector_from_name(selector_name)
        message_hash = hash_message(
            account.contract_address, recipient_address, selector, calldata, nonce)
        sig_r, sig_s = self.sign(message_hash)

        return await account.execute(recipient_address, selector, calldata, nonce).invoke(signature=[sig_r, sig_s])


def hash_message(sender, recipient_address, selector, calldata, nonce):
    message = [
        sender,
        recipient_address,
        selector,
        compute_hash_on_elements(calldata),
        nonce
    ]
    return compute_hash_on_elements(message)
